lib/wallet/src/Cardano/Wallet.hs

```diff
@@ -2569,141 +2580,213 @@ buildAndSignTransactionNew
        , IsOwned s k ktype
        , IsOurs s RewardAccount
        , GenChange s
+       , AddressBookIso s
        )
-    => Tracer IO WalletLog
-    -> TimeInterpreter (Either PastHorizonException)
+    => TimeInterpreter (ExceptT PastHorizonException IO)
     -> DBLayer IO s k
     -> NetworkLayer IO Block
     -> TransactionLayer k ktype SealedTx
+    -> Passphrase "user"
     -> WalletId
     -> ArgGenChange s
     -> AnyRecentEra
+    -> PreSelection
+    -> TransactionCtx
+    -> IO (BuiltTx, UTCTime)
+buildAndSignTransactionNew ti db@DBLayer{..} netLayer txLayer pwd walletId
+    argGenChange era preSelection txCtx = do
+    --
+    stdGen <- initStdGen
+    pureTimeInterpreter <- snapshot ti
+    protocolParameters <- currentProtocolParameters netLayer
+    let wrapError = ExceptionWitnessTx . ErrWitnessTxWithRootKey
+    throwOnError <=< runExceptT $ withRootKey db walletId pwd wrapError $
+        \rootKey scheme -> lift $ atomically $ do
+            pendingTxs <- fmap fromTransactionInfo <$>
+                readTransactions
+                    walletId Nothing Descending wholeRange (Just Pending)
+            txBuiltForSlot <- modifyDBMaybe walletsDB $ do
+                let err = ExceptionConstructTx . ErrConstructTxNoSuchWallet
+                adjustNoSuchWallet walletId err $ \(s :: WalletState s) -> do
+                    let wallet = WalletState.getLatest s
+                    let (errorOrBuiltTx, wallet') =
+                            buildAndSignTransactionPure @k @ktype @s @n
+                                nullTracer
+                                pureTimeInterpreter
+                                (Set.fromList pendingTxs)
+                                rootKey
+                                scheme
+                                pwd
+                                protocolParameters
+                                txLayer
+                                argGenChange
+                                era
+                                preSelection
+                                txCtx
+                            & runExceptT . withExceptT
+                                (either ExceptionBalanceTx ExceptionConstructTx)
```

Comment:
If we get an error, the wallet state hasn't been updated, and writing it to DB is pointless. Should we not run this `ExceptT` after `runStateT` instead?

Reply:
Agree, order changed.